### PR TITLE
Backport hash agg commits from REL_13_STABLE

### PR DESCRIPTION
**Pull request by adam8157**
Greenplum switched to the upstream implementation of hash agg spilling,
as a result, the nodeAgg.c code is more similar to the v13 version than v12.

This PR backports related fix commits from REL_13_STABLE.

### CODE REVIEW

**Comment by ashwinstar** [comment]
Thanks @adam8157 for completing this aspect.

File: src/backend/utils/sort/logtape.c [hunk 1]
```diff
@@ -1271,9 +1273,24 @@ LogicalTapeTell(LogicalTapeSet *lts, int tapenum,
...
-	return lts->nBlocksAllocated - lts->nHoleBlocks;
+#ifdef USE_ASSERT_CHECKING
+	for (int i = 0; i < lts->nTapes; i++)
+	{
+		LogicalTape *lt = &lts->tapes[i];
+		/*
+		 * Greenplum interrupts the sort and set QueryFinishPending to true on
+		 * purpose in the test query_finish_pending.
+		 */
+		if (!QueryFinishPending)
```
Comment (adam8157) [inline src/backend/utils/sort/logtape.c:1291]: @yaowangm could you please help review this? Thanks.

File: src/backend/executor/nodeAgg.c [hunk 2]
```diff
@@ -2202,21 +2160,46 @@ static void
...
+		TupleHashTable hashtable = perhash->hashtable;
+		TupleTableSlot *hashslot = perhash->hashslot;
+		TupleHashEntry entry;
+		uint32		hash;
+		bool		isnew = false;
+		bool	   *p_isnew;
+
+		/* if hash table already spilled, don't create new entries */
+		/* or the hash table is filled on the streaming mode */
+		p_isnew = (aggstate->hash_spill_mode || (aggstate->streaming && aggstate->table_filled)) ? NULL : &isnew;
```
[the remark this answers is not in the document]
Comment (adam8157) [reply src/backend/executor/nodeAgg.c:2202]: Correct, it's for streaming, I moved the logic to this function just as how the upstream refactors.

**Review by yaowangm — APPROVED** [review]
LGTM. Thanks for your work.

**Pull request merged**